src/TwilightImperium.lua

```diff
@@ -0,0 +1,131 @@
+-- TwilightImperium.lua
+-- Firoso 2019
+-- Written for use by the Twilight Imperium Tabletop Simulator Community
+-- https://github.com/TwilightImperiumContentCreators/TTS-TwilightImperium
+
+-- This file contains classes, functions, and tables for the domain models
+-- for Twilight Imperium
+
+-- STATIC DEFINITIONS --
+Constants = {
+  PlanetTypes = {
+    None = "None",
+    Hazardous = "Hazardous",
+    Cultural = "Cultural",
+    Industrial = "Industrial",
+    MecatolRex = "Mecatol Rex",
+    HomeWorlds =
+    {
+      Sol = "Sol HomeWorld",
+      -- etc
+    },
+    TechnologySpeciality =
+    {
+      None = "",
+      Red = "Red - Warfare",
+      Blue = "Blue - Propulsion",
+      Yellow = "Yellow - Cybernetic",
+      Green = "Green - Biotic"
+    }
+  },
+  Anomolies = {
+    AlphaWormhole = "ɑ Wormhole",
+    BetaWormhole = "ᵝ Wormhole",
+    DeltaWormhole = "δ Wormhole",
+    AsteroidField = "Asteroid Field",
+    Supernova = "Supernova",
+    Nebula = "Nebula",
+    GravityRift = "GravityRift"
+  }
+}
+
+-- ABSTRACTIONS --
+-- the class 'Exhaustable' used to represent assets that can be
+-- exhausted to gain some benefit.
+function Exhaustable()
+    -- public members
+    local self = {
+      -- indicates if this asset is exhausted or not
+      exhausted = false,
+    }
+
+    function self.exhaust()
+      if exhausted then
+        error ("Already exhausted", 2)
```

Comment:
`error` is something of an anti-pattern here - As `exhausted` is public, `exhaust` will be used as a guard to behave differently if the resource was already exhausted. Raising an error will mean that the caller will have to manually catch it with `pcall` -

```lua
if pcall(tile.exhaust()) then
  print("Youre already out!")
end
```

Instead, it is lua convention (again, supported by the stdlib) to return a falsy value and a reason on failure.

```lua
function Exhaustable:exhaust()
  if self.exhausted then
    return false, "Already exhausted"
  end
  self.exhausted = false
  return true
end

if tile.exhaust() then
  print("Youre already out!")
end
```

Or, in case the caller does consider it fatal, the `assert` function can be used to raise the `message` on failure. More examples can be found [here](https://www.lua.org/pil/8.3.html)

Reply:
Excellent feedback!